### PR TITLE
RavenDB-19992: Low level optimization of key encoding process

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-19992

### Additional description
Assembler level optimization to achieve faster encoding of keys. 

### Type of change
- Optimization

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- No

### CODE REVIEW

**Comment by redknightlois** [comment]
image

File: src/Voron/Data/CompactTrees/PersistentDictionaryLocator.cs [hunk 1]
```diff
@@ -65,14 +55,33 @@ public void Set(long pageNumber, PersistentDictionary dictionary)
...
+                case null:
+                    _cache[position] = dictionary;
+                    break;
+                case PersistentDictionary existing:
+                    if (existing.DictionaryId == dictionary.DictionaryId)
+                        return; // should never happen
+                    _cache[position] = new Dictionary<long, PersistentDictionary>
+                    {
+                        [existing.DictionaryId] = existing,
+                        [dictionary.DictionaryId] = dictionary
```
Comment (ayende) [inline src/Voron/Data/CompactTrees/PersistentDictionaryLocator.cs:75]: Subtle behavior here. 
If we have two dictionaries that are on the _same_ id, we'll use the `dictionary` one, not the `existing`, it is not an error in this case.

**Comment by redknightlois** [comment]
@arekpalinski I solved both issues. For RavenDB-20142 I found and solved it. However, for RavenDB-20143 I decided to just get rid of the compact key sharing. With the current usage pattern the contribution of allocations of the compact keys is much smaller than it was initially, therefore, while I didnt remove the infrastructure it is safe to remove this ability until we can investigate deeply why and how sharing compact keys instances corrupt the underlying storage.

**Comment by arekpalinski** [comment]
@redknightlois Please check @ayende's comment. I'm merging it and will verify the fix.

**Pull request merged**